internal/download/download.go

```diff
@@ -0,0 +1,311 @@
+}
+
+type versionEntry struct {
+	name    string
+	version string
+}
+
+type hashEntry struct {
+	hash string
+	file string
+	url  *url.URL
+}
+
+// MustLoadChecksums loads a file containing checksums.
+func MustLoadChecksums(file string) *ChecksumDB {
+	content, err := os.ReadFile(file)
+	if err != nil {
+		log.Fatal("can't load checksum file: " + err.Error())
+	}
+	db, err := ParseChecksums(content)
+	if err != nil {
+		log.Fatalf("invalid checksums in %s: %v", file, err)
+	}
+	return db
+}
+
+// ParseChecksums parses a checksum database.
+func ParseChecksums(input []byte) (*ChecksumDB, error) {
+	var (
+		csdb    = new(ChecksumDB)
+		rd      = bytes.NewBuffer(input)
+		lastURL *url.URL
+	)
+	for lineNum := 1; ; lineNum++ {
+		line, err := rd.ReadString('\n')
+		if err == io.EOF {
+			break
+		}
+		line = strings.TrimSpace(line)
+		switch {
+		case line == "":
+			// Blank lines are allowed, and they reset the current urlEntry.
+			lastURL = nil
+
+		case strings.HasPrefix(line, "#"):
+			// It's a comment. Some comments have special meaning.
+			content := strings.TrimLeft(line, "# ")
+			switch {
+			case strings.HasPrefix(content, "version:"):
+				// Version comments define the version of a tool.
+				v := strings.Split(content, ":")[1]
+				parts := strings.Split(v, " ")
+				if len(parts) != 2 {
+					return nil, fmt.Errorf("line %d: invalid version string: %q", lineNum, v)
+				}
+				csdb.versions = append(csdb.versions, versionEntry{parts[0], parts[1]})
+
+			case strings.HasPrefix(content, "https://") || strings.HasPrefix(content, "http://"):
+				// URL comments define the URL where the following files are found. Here
+				// we keep track of the last found urlEntry and attach it to each file later.
+				u, err := url.Parse(content)
+				if err != nil {
+					return nil, fmt.Errorf("line %d: invalid URL: %v", lineNum, err)
+				}
+				lastURL = u
+			}
+
+		default:
+			// It's a file hash entry.
+			fields := strings.Fields(line)
+			if len(fields) != 2 {
+				return nil, fmt.Errorf("line %d: invalid number of space-separated fields (%d)", lineNum, len(fields))
+			}
+			csdb.hashes = append(csdb.hashes, hashEntry{fields[0], fields[1], lastURL})
+		}
+	}
+	return csdb, nil
+}
+
+// Files returns an iterator over all file names.
+func (db *ChecksumDB) Files() iter.Seq[string] {
+	return func(yield func(string) bool) {
+		for _, e := range db.hashes {
+			if !yield(e.file) {
+				return
+			}
+		}
+	}
+}
+
+// Verify checks whether the given file is valid according to the checksum database.
+func (db *ChecksumDB) Verify(path string) error {
+	basename := filepath.Base(path)
+	hash := db.findHash(basename)
+	if hash == "" {
+		return fmt.Errorf("no known hash for file %q", basename)
+	}
+	return verifyHash(path, hash)
+}
+
+// VerifyAll downloads all files and checks that they match the checksum given in
+// the database. This task can be used to sanity-check new checksums.
+func (db *ChecksumDB) VerifyAll() {
```

Comment:
```suggestion
func (db *ChecksumDB) DownloadAndVerifyAll() {
```
What do you think about something like this? `VerifyAll()` doesn't really convey to me that it will actually download all the files, especially given `Verify()` does not download anything.